### ISSUE TITLE
Analyzer can report `invalid_null_aware_operator` _and_ follow-on errors from the resultant nullable type

### ISSUE THREAD

**Issue by srawlins** [dart-lang/sdk#49190]
See this code:

```dart
bool f(int i) => i?.isEven;
```

Analyzer produces one error:

> A value of type 'bool?' can't be returned from the function 'f' because it has a return type of 'bool'.

and one warning:

> The receiver can't be null, so the null-aware operator '?.' is unnecessary.

This is a bummer, because the "real" issue is the warning, and the error is sort of secondary; the error is indirectly solved by directly solving the warning.

Can we avoid the double reporting? Should we? Given that these are both specified by the null safety spec (I think), _must_ we report each? In other words, are there situations/users/codebases where the warning may be suppressed or ignored, and so we must continue to report the error? It might be very complicated to detect that `invalid_null_aware_operator` has been reported (and not suppressed) in order to _not_ report `argument_type_not_assignable` (or other similar assignability errors).

CC @leafpetersen @eernstg for any comments from the language team.

**Comment by srawlins** [comment]
Oh and this is motivated by our internal build system, which groups errors on the same line into one report. E.g.:

```none
291: foo.bar?.type,                    The argument type 'SomethingPB_Enum?' can't be assigned to the parameter type
                                       'SomethingPB_Enum'. (warning) The receiver can't be null, so the null-aware operator '?.' is
                                       unnecessary. #argument_type_not_assignable #invalid_null_aware_operator
```

Note that the more useful diagnostic is reported second. Until I've read the second diagnostic, it looks to me like a real puzzle to get a non-null value from an expression with a null-aware property access.

**Comment by bwilkerson** [comment]
Baring any objections from the language team, I would say that we ought to recover from the warning by not reporting the error (treat a null-aware operator on a non-nullable target as if the non-null-aware version of the operator had been used). I think that's what's best for users because, as you said, it points them to the real issue.

**Comment by eernstg** [comment]
I agree that the diagnostic about `?.` being unnecessary addresses the real issue directly, but I tend to think about the other issue ("can't return a `bool?`") as an independent consequence. We could have any number of diagnostics based on the fact that an expression has an unintended/suboptimal type (with `var x = e?.foo();`, many usages of `x` could fail in various ways), and there is no special connection between "unnecessary `?.`" and "not assignable" as such. So we're probably looking for something which is a bit more general.

In particular, I don't think it is a sufficiently general strategy to treat an unnecessary `?.` as if it had been a plain `.`. Another danger with that approach is that it might be confusing: If the developer corrects the non-nullable receiver expression to something nullable (so the `?.` wasn't intended to be unnecessary, we just had a typo in the receiver) then the "not assignable" error would emerge out of thin air. And how would we handle the situation where warnings are not shown at all?

The core topic here seems to be dependencies among diagnostic messages.

We do eliminate most error dependencies today, simply by not reporting errors with dependencies:

```dart
void main() {
  UnknownClass u = 'A string'; // No diagnostics on the initializing expression.
  u.toString(42); // No diagnostics here.
}
```

We get an error on the usage of an undefined name as a type, but no error on the initialization, and no error on the subsequent use of the variable (and it's _not_ just treated as `var u = 'A string';`). I can't immediately find a counterexample to the assumption that `u` is treated as if it had the type `dynamic`, but it could also be a special `ErrorType` (which is treated similarly to `dynamic`, such that we don't report any errors/warnings).

If we had been discussing the relationship between two _errors_ here, we'd simply get rid of the dependent error by letting `i?.isEven` have the type `ErrorType`. We do not test diagnostic dependencies in language tests, and we don't specify `ErrorType` and how to handle it (or anything like that). So there would not have been anything new or controversial in letting `i?.isEven` have the type `ErrorType` _if_ the use of `?.` had been an error.

However, the use of `?.` in this situation is a warning, and, presumably, we don't want to "let a warning cancel an error", and we still have to define how to deal with the situation where warnings are suppressed.

Maybe we could prioritize the diagnostic messages in this situation? Show independent diagnostic messages before dependent ones (in IDEs and on the command line)? Include some kind of a link from the dependent ones to their dependencies?

**Comment by srawlins** [comment]
> I can't immediately find a counterexample to the assumption that u is treated as if it had the type dynamic, but it could also be a special ErrorType (which is treated similarly to dynamic, such that we don't report any errors/warnings).

This is correct; we just set the static type, in analyzer, to dynamic. This is a pain point in some circumstances, like in code gen, when the code generator would like to know that the code had an error, rather than finding an object of type `dynamic`.

> However, the use of `?.` in this situation is a warning, and, presumably, we don't want to "let a warning cancel an error", and we still have to define how to deal with the situation where warnings are suppressed.

Yeah this was my main worry.

> Maybe we could prioritize the diagnostic messages in this situation? Show independent diagnostic messages before dependent ones (in IDEs and on the command line)? Include some kind of a link from the dependent ones to their dependencies?

Interesting ideas. @bwilkerson would know more about what we can do in IDEs, which is the primary location where we show errors. And what is expected etc. We may not have control over the order that errors appear in the Problems panel. The case for the CLI and our internal builds are probably enough under our control that we could order issues on the same line intentionally.

**Comment by bwilkerson** [comment]
> In particular, I don't think it is a sufficiently general strategy to treat an unnecessary `?.` as if it had been a plain `.`.

I agree that consistency is important, both in language design and in the UX. One of the recovery techniques that we consistently apply is to silence follow-on diagnostics by treating the code as if it had been written correctly. I see treating `?.` as if it had been a `.` to be just another application of the more general technique.

> Another danger with that approach is that it might be confusing: If the developer corrects the non-nullable receiver expression to something nullable (so the ?. wasn't intended to be unnecessary, we just had a typo in the receiver) then the "not assignable" error would emerge out of thin air.

I don't think users would perceive the diagnostic as emerging out of thin air; I think it would be perceived as the result of modifying the code.

> And how would we handle the situation where warnings are not shown at all?

We'd need to be a but smarter about warnings and produce the error only in cases where the warning would not be produced.

> > I can't immediately find a counterexample to the assumption that `u` is treated as if it had the type `dynamic`, but it could also be a special `ErrorType` (which is treated similarly to `dynamic`, such that we don't report any errors/warnings).
>
> This is correct; we just set the static type, in analyzer, to `dynamic`. This is a pain point in some circumstances, like in code gen, when the code generator would like to know that the code had an error, rather than finding an object of type `dynamic`.

I have no conceptual issue with defining an `ErrorType`, but it might be the case that it would need to be understood by type inference, which might have an impact on shared code, and hence the CFE. I don't know that that's the case, but I'd want us to investigate the possible impact of such a change before making it.

> ... presumably, we don't want to "let a warning cancel an error" ...

I agree that that's a concern, and possibly a bit enough concern to require a different solution to this issue.

> Maybe we could prioritize the diagnostic messages in this situation? Show independent diagnostic messages before dependent ones (in IDEs and on the command line)? Include some kind of a link from the dependent ones to their dependencies?

We don't really have any formal definition of 'dependent' diagnostics, nor any representation of such a concept in the implementation. What we do have available today is the severity of the diagnostic, and it would certainly be reasonable, in places where we can control the presentation order, to display higher severity ('error') diagnostics before lower severity ('warning') diagnostics.

> We may not have control over the order that errors appear in the Problems panel.

That is correct. IDEs typically provide the user with some level of control over the presentation, which means that we're pretty much limited to controlling the data that is presented.

-----

I'll offer an alternative solution. I haven't investigated it yet, but it's possible that we could extend our analysis when we detect the error (return_of_invalid_type and possibly others) to notice that the two types differ only by nullability suffix and that the nullability was introduced by the `?.` operator. Under those conditions we could produce a message (some combination of problem, correction, or context) that suggests that the operator might be the underlying problem. We could then suppress the warning (possibly by leaving a breadcrumb when producing the error) without loosing any information (from the user's perspective).
